modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/database/H2TreeInlineObjectDetector.java

```diff
@@ -68,26 +85,61 @@
                 continue;
             }
 
-            if (r.getValue(ih.columnIndex()) == ValueNull.INSTANCE)
+            Value val = r.getValue(ih.columnIndex());
+
+            if (val == ValueNull.INSTANCE)
                 return false;
 
             int type = PageUtils.getByte(pageAddr, off + fieldOff);
 
-            if (type == 0) {
-                inlineObjSupported = false;
+            //We can have garbage in memory and need to compare data.
+            if (type == Value.JAVA_OBJECT) {
+                int len = PageUtils.getShort(pageAddr, off + fieldOff + 1);
 
-                return true;
-            }
-            else if (type == Value.JAVA_OBJECT) {
-                inlineObjSupported = true;
+                boolean isFull = (len & 0x8000) == 0;
+
+                len = len & 0x7FFF;
+
+                byte[] originalObjBytes = val.getBytesNoCopy();
+
+                // read size more then available space or more then origin length
+                if(len > inlineSize - fieldOff - 3 || len > originalObjBytes.length){
```

Comment:
`len > originalObjBytes.length` seems to mean unexpected (garbage) data in current position. Cannot we do comparison in branch `if (type == Value.JAVA_OBJECT)` using following strategy:
1. Prepare (virtually) an expected byte content which is saved by regular inline procedures.
2. Compare that content with bytes stored in the index page.
3. Exact match means that JAVA_OBJECT inline is (almost 100%) supported. Otherwise it is not (tree corruption is possible, might be warning is a good idea). In both cases we can finish tree scan here.

Also using of `InlineIndexHelper.compare` is dangerous below, because it is more broad method and it is used in different places. But here we are just solving a compatibility issue from certain point in history.

Reply:
for your comparison proposal - by fact it do the same way.

objections to use InlineIndexHelper.compare  don't understand. Let's discuss internally.